### PR TITLE
DEV: Proof of concept – Refactor mentions notifications logic

### PR DESCRIPTION
**Pull request by AndrewPrigorshnev**
This PR contains a complete rewrite of the chat mention notifications logic. I've removed the old `Notifier` and `NotifyMentioned` classes and wrote a new implementation that consists only of the `NotifyMentioned` job. I reused big chunks of the old code along the way, but refactored and regrouped them.

You'll see many changes to tests in the diff, but the idea was to touch them as less as possible and to let them keep testing the same things. In order to leave the tests but make the refactoring possible, I moved them one level upper and made them call the  `ProcessMessage` job (which called the former `Notifier` and `NotifyMentioned` job). Then I had to go through all tests and correct their setups to made sure there were no false positives.

So you can assume all the old tests are still there, with these two exceptions:
1. I [dropped](https://github.com/discourse/discourse/pull/25844/commits/861c7663b1bee833a1c8f243ad961b89b2c34cc3) a test that tested solely details of interaction between `Notifier` and the old `NotifyMentioned` job
2. I dropped two tests from the `ParsedMentions` class, because I moved the related logic into the `NotifyMentioned` job. That logic is tested in other tests, you can see all that in [this commit](https://github.com/discourse/discourse/pull/25844/commits/e88fb3b002f03166606081a2168c4ce58f6f3217)

As I've mentioned already, the old implementation consisted of two classes: `Notifier` and the `NotifyMentioned`  job. `Notifier` looked at mentions parsed from a message and prepared a hash of users who should be notified. Then `Notifier` passed that hash to the `NotifyMentioned` job, and the job looked for related `chat_mentions` records in the database and sent notifications.

The new version consists only of the `NotifiyMentioned` job, the job takes `message_id` as an argument, loads `chat_mention` records for the message right away and sends notifications. The new implementation is still idempotent and doesn't send duplicates of notifications in case the job was called twice or when editing a previously created message.

Apart from that, in the new implementation:
1. There is the only code path for creating and editing messages (we had separate `notify_new` and `notify_edit` methods before)
1. Some domain logic moved from `ParsedMentions` and `NotifyMentioned` into models
1. The code that sends mentions warnings is grouped in a specialized class – https://github.com/discourse/discourse/blob/9607ee1235e8dc4bd004da69f35a71bcbb0065ff/plugins/chat/app/services/chat/mentions_warnings.rb.

The work in this PR is not fully finished yet, I'm opening it for review so we can decide if we want to apply this refactoring.  My goal was to bring it to the state when all backend tests pass and there are no problems when smoke testing in the UI. There are still some details that need to be moved to the new implementation (and tested), but I feel I've made at least 80% of all work already.

If we decide to apply this refactoring, I would go in three steps:
1. Extract changes to tests from this PR and merge them first (since they test the same behavior as before, they should pass against the current main)
2. Add more tests. Message editing definitely needs better coverage, I feel we also may add more tests for mention warnings.
3. Check performance, address N + 1's and fix remaining little details that haven't been moved from the old implementation yet

There is some more space for improvements in this code, for example I feel even more logic can be extracted from the `ParsedMentions` class, but I tried to not increase the surface of the refactoring too much.

There are some errors in system specs, I'm currently working on fixing them.

### CODE REVIEW

File: app/models/user.rb [hunk 1]
```diff
@@ -620,10 +620,23 @@ def ignored_user_ids
...
+
   def muted_user_ids
     @muted_user_ids ||= muted_users.pluck(:id)
   end
 
+  def mutes?(user)
+    muted_user_ids.include?(user.id)
+  end
+
+  def doesnt_want_to_hear_from(user)
```
Comment (jjaffeux) [inline app/models/user.rb:635]: I would suggest `disregards(user)` or `ignores_or_mutes?(user)` instead

File: app/services/post_alerter.rb [hunk 2]
```diff
@@ -48,6 +48,10 @@ def self.create_notification_alert(user:, post:, notification_type:, excerpt: ni
     end
   end
 
+  def self.desktop_notification(user, payload)
+    ::MessageBus.publish("/chat/notification-alert/#{user.id}", payload, user_ids: [user.id])
```
Comment (jjaffeux) [inline app/services/post_alerter.rb:52]: I know it's draft, but just noting we should have "chat" in this code

File: plugins/chat/app/jobs/regular/chat/notify_mentioned.rb [hunk 3]
```diff
@@ -3,184 +3,98 @@
...
+        user_ids =
+          @message
+            .chat_mentions
+            .joins(
+              "LEFT OUTER JOIN chat_mention_notifications cmn ON cmn.chat_mention_id = chat_mentions.id",
+            )
+            .joins("LEFT OUTER JOIN notifications n ON cmn.notification_id = n.id")
+            .where.not("n.user_id IS NULL")
+            .pluck("n.user_id")
+        Set.new(user_ids)
```
Comment (ZogStriP) [inline plugins/chat/app/jobs/regular/chat/notify_mentioned.rb:38]: I believe you can do `.to_set` on the collection to create a set

File: plugins/chat/app/jobs/regular/chat/notify_mentioned.rb [hunk 4]
```diff
@@ -3,184 +3,100 @@
 module Jobs
```
Comment (jjaffeux) [inline plugins/chat/app/jobs/regular/chat/notify_mentioned.rb:3]: Ideally I would prefer Jobs to be almost logicless, I see them as controllers. We should maybe consider moving all this code to services.

**Pull request closed without merging**